README.md
# A simple PHP & SQLite Login Script #

Heavily modified version. :3

CHANGELOG.md
# CHANGELOG #

May 22, 2013
* 1-minimal: Converted to use PDO and SQLite, very messy but it works correctly.

May 18, 2013
* 1-minimal: html5 form attributs that (optionally) validate the input fields on client's browsers:
* 1-minimal: min/max length for email input fields
* 1-minimal: min/max length for username input field, additionally html5 string check (a-z, A-Z, 0-9)
* 1-minimal: user need to provide email now, registration without email is not possible anymore
* 1-minimal: PHP checks for username structure (a-z, A-Z, 0-9), email structure
* 1-minimal: removed 64 char limit for password. passwords can now be 1024 chars
* 1-minimal: login.sql (in "_install" folder) renamed to users.sql (as it is the name of the database table)

May 12, 2013
* changed hashing algorithm from blowfish/SHA256 to SHA512
* changed database creation files (due to new SHA512 hashing algorithm)
* changed database column "user_password_hash" from CHAR(60) to CHAR(118) [as hash is always 118 chars long]
* added HTML5 attributes to views (type="email", required etc.)

April 26, 2013
* complete makeover, nearly all files have been touched
* registration process is now in seperate class and seperate init file / view (register.php etc.)
* massive reduction of the views: no css, no js, no unnecessary stuff. just pure naked basics
* entire project is now free of php "notice" error when you have hard error reporting
* documented nearly EVERYTHING
* entire project tries to be PSR-1/2 compilant, which means: everything fits to the PSR coding standards
* (see https://github.com/php-fig/fig-standards for more)
* changed database column "user_password_hash" from TEXT to CHAR(60) [as hash is always 60 chars long]
* changed database column "user_email" from TEXT to VARCHAR(64) [variable length string, up to 64 chars]

1-minimal/index.php
<?php
require 'account.php';
$account = new Account();
if ($account->isLoggingIn()) {
	//prints the return status of the login
	print_r($account->login());
}
if ($account->isRegistering()) {
	//prints the return status of the register
	print_r($account->register());
}
if ($account->isLoggedIn()) {
	echo '<h3>Welcome ' . $account->getUsername() . '</h3>';
	echo '<a href="index.php?logout">Logout</a>';
} else {
?>
	<h3>Login</h3>
	<form method="post" action="index.php">
		<input type="text" placeholder="username" name="username"><br>
		<input type="password" placeholder="password" name="password"><br><br>
		<input type="submit" name="login" value="Log In">
	</form>
	<h3>Register</h3>
	<form method="post" action="index.php">
		<input type="text" placeholder="username" name="username"><br>
		<input type="password" placeholder="password" name="password"><br>
		<input type="password" placeholder="password again" name="password_repeat"><br><br>
		<input type="submit" name="register" value="Register">
	</form>
<?php
}

1-minimal/account.php
<?php
/** name of the database file */
define('DB_FILENAME', 'account.db');
/** schema file */
define('SCHEMA_FILENAME', 'schema.sql');
/* TODO: Allow define of POST, GET, and SESSION variables */
/* TODO: Allow registration to be disabled */
/**
 * Handles login, logout, registration and session.
 */
class Account {
    private $database  = null; // database connection
    private $logged_in = false; // if the user is logged in
    private $username  = null; // username
    const SUCCESS         = false;
    const BAD_USERNAME    = 1;
    const BAD_PASSWORD    = 2;
    const DATABASE_ERROR  = 3;
    const INCOMPLETE_DATA = 4;
    const USERNAME_TAKEN  = 5;
    /**
     * the function "__construct()" automatically starts whenever an object of this class is created,
     * you know, when you do "$account = new Account();"
     */
    public function __construct() {
        session_start();
        // if user tried to log out
        if (isset($_GET['logout'])) {
            $this->Logout();
        }
        // if user has an active session on the server
        elseif (!empty($_SESSION['username']) && ($_SESSION['logged_in'] == 1)) {
            $this->loginSession();
        }
    }
    private function openDatabase() {
        try {
            $create_db = false;
            if (!file_exists(DB_FILENAME)) {
                $create_db = true;
            }
            $this->database = new PDO('sqlite:' . DB_FILENAME);
            if ($create_db) {
                $this->database->exec(file_get_contents(SCHEMA_FILENAME));
            }
            $this->database->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
            $this->database->setAttribute(PDO::ATTR_EMULATE_PREPARES, false);
        }
        catch (PDOException $e) {
            return false;
        }
        return true;
    }
    private function loginSession() {
        $this->logged_in = true;
        $this->username = $_SESSION['username'];
    }
    /**
     * Returns true if there is a form login attempt pending. Call login to process this.
     */
    public function isLoggingIn() {
        return isset($_POST["login"]);
    }
    /**
     * Performs a pending login, returns false on success, otherwise returns a error code.
     */
    public function login() {
        if (empty($_POST['username']) || empty($_POST['password']))
            return self::INCOMPLETE_DATA;
        //TODO throw incomplete login data error
        if ($this->openDatabase()) {
            //lookup user in the database
            $sql = 'SELECT username, password_hash FROM users WHERE username = ?;';
            $sql = $this->database->prepare($sql);
            $sql->execute(array(
                $_POST['username']
            ));
            // if we can get a result object, then the user exists
            if ($result = $sql->fetchObject()) {
                //check the password
                if (crypt($_POST['password'], $result->password_hash) == $result->password_hash) {
                    // write user data into a session
                    $_SESSION['username'] = $result->username;
                    $_SESSION['logged_in'] = true;
                    $this->username = $result->username;
                    $this->logged_in = true;
                } else {
                    return self::BAD_PASSWORD;
                }
            } else {
                return self::BAD_USERNAME;
            }
            $this->database = null;
        } else {
            return self::DATABASE_ERROR;
        }
        return self::SUCCESS;
    }
    /**
     * Returns true if there is a registration attempt pending. Call register to process this.
     */
    public function isRegistering() {
        return isset($_POST["register"]);
    }
    /**
     * Performs a pending registration, returns false on success, otherwise returns a error code.
     */
    public function register() {
        if (empty($_POST['username']))
            return self::INCOMPLETE_DATA;
        if (empty($_POST['password']) || empty($_POST['password_repeat']))
            return self::INCOMPLETE_DATA;
        if ($_POST['password'] !== $_POST['password_repeat'])
            return self::BAD_PASSWORD;
        if (strlen($_POST['username']) > 64)
            return self::BAD_USERNAME;
        //alphanumeric and underscore   3 to 64 characters
        if (!preg_match('/^[_a-zA-Z0-9]{3,64}$/', $_POST['username']))
            return self::BAD_USERNAME;
        if ($this->openDatabase()) {
            //truncate password to 128 characters, anything more is a bit silly imho
            $password = substr($_POST['password'], 0, 128);
            //create a random hex string for our salt
            $salt = unpack('H*', mcrypt_create_iv(CRYPT_SALT_LENGTH >> 1, MCRYPT_DEV_URANDOM))[1];
            //SHA-512
            $algorithm = '$6$rounds=5000$';
            //create our hash
            $hash = crypt($password, $algorithm . $salt);
            $sql = 'SELECT * FROM users WHERE username = ?;';
            $sql = $this->database->prepare($sql);
            $sql->execute(array(
                $_POST['username']
            ));
            if ($sql->fetch(PDO::FETCH_NUM) === false) {
                //username is not taken, write new data
                $sql = 'INSERT INTO users (username, password_hash) VALUES(?, ?);';
                $sql = $this->database->prepare($sql);
                $sql->execute(array(
                    $_POST['username'],
                    $hash
                ));
                if ($sql->rowCount() != 1) {
                    //account creation failure
                    return self::DATABASE_ERROR;
                }
            } else {
                return self::USERNAME_TAKEN;
            }
            $this->database = null;
        } else {
            return self::DATABASE_ERROR;
        }
        return self::SUCCESS;
    }
    /**
     * Perform a logout
     */
    public function logout() {
        $_SESSION = array();
        session_destroy();
        $this->logged_in = false;
    }
    /**
     * Returns true if a user is currently logged in.
     */
    public function isLoggedIn() {
        return $this->logged_in;
    }
    /**
     * Returns the username of the currently logged in user, null otherwise.
     */
    public function getUsername() {
        return $this->username;
    }
}

1-minimal/db_login.php
<?php

/**
 * class Login
 * handles the user login/logout/session
 * 
 * @author Panique <panique@web.de>
 * @version 1.2
 */
class Login {

    private     $db_connection              = null;                     // database connection
    
    private     $user_name                  = "";                       // user's name
    private     $user_email                 = "";                       // user's email
    private     $user_password_hash         = "";                       // user's hashed and salted password
    private     $user_is_logged_in          = false;                    // status of login

    public      $errors                     = array();                  // collection of error messages
    public      $messages                   = array();                  // collection of success / neutral messages
    
    
    /**
     * the function "__construct()" automatically starts whenever an object of this class is created,
     * you know, when you do "$login = new Login();"
     */    
    public function __construct() {
        
        // create/read session
        session_start();                                        

        // check the possible login actions:
        // 1. logout (happen when user clicks logout button)
        // 2. login via session data (happens each time user opens a page on your php project AFTER he has sucessfully logged in via the login form)
        // 3. login via post data, which means simply logging in via the login form. after the user has submit his login/password successfully, his
        //    logged-in-status is written into his session data on the server. this is the typical behaviour of common login scripts.
		
        // if user tried to log out
        if (isset($_GET["logout"])) {

            $this->doLogout();
                    
        } 
        // if user has an active session on the server
        elseif (!empty($_SESSION['user_name']) && ($_SESSION['user_logged_in'] == 1)) {

            $this->loginWithSessionData();                

        // if user just submitted a login form
        } elseif (isset($_POST["login"])) {

                $this->loginWithPostData();
                
        }        
    }    
    

    private function loginWithSessionData() {
        
        // set logged in status to true, because we just checked for this:
        // !empty($_SESSION['user_name']) && ($_SESSION['user_logged_in'] == 1)
        // when we called this method (in the constructor)
        $this->user_is_logged_in = true;
        
    }
    

    private function loginWithPostData() {
        
        // if POST data (from login form) contains non-empty user_name and non-empty user_password
        if (!empty($_POST['user_name']) && !empty($_POST['user_password'])) {
            
            // create a database connection, using the constants from db_config.php (which we loaded in index.php)
			try {
				$create_db = false;
				
				if(!file_exists(DB_FILENAME))
					$create_db = true;
					
				$this->db_connection = new PDO('sqlite:'.DB_FILENAME);
				
				if($create_db)
					$this->db_connection->exec(file_get_contents('schema.sql'));
				
				$this->db_connection->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
				
				$this->db_connection->setAttribute(PDO::ATTR_EMULATE_PREPARES, false);
				
			} catch (PDOException $e) {
			
				$this->errors[] = "Database connection problem.";
				
				return;
				
			}
            
            // if no connection errors (= working database connection)
		
			// database query, getting all the info of the selected user
			$statement = $this->db_connection->prepare('SELECT user_name, user_email, user_password_hash FROM users WHERE user_name = :username;');

			$statement->execute(array('username' => $_POST['user_name']));
			
			// if this user exists
			// get result row (as an object)
			if ( $result_row = $statement->fetchObject() ) {

				// using PHP's crypt function to 
				// this is currently (afaik) the best way to check passwords in login processes with PHP/SQL
				if (crypt($_POST['user_password'], $result_row->user_password_hash) == $result_row->user_password_hash) {

					// write user data into PHP SESSION [a file on your server]
					$_SESSION['user_name'] = $result_row->user_name;
					$_SESSION['user_email'] = $result_row->user_email;
					$_SESSION['user_logged_in'] = 1;

					// set the login status to true
					$this->user_is_logged_in = true; 

				} else {

					$this->errors[] = "Wrong password. Try again.";

				}                

			} else {

				$this->errors[] = "This user does not exist.";
			}
            
        } elseif (empty($_POST['user_name'])) {

            $this->errors[] = "Username field was empty.";

        } elseif (empty($_POST['user_password'])) {

            $this->errors[] = "Password field was empty.";
        }           
        
    }
    
    /**
     * perform the logout
     */
    public function doLogout() {
            
            $_SESSION = array();
            session_destroy();
            $this->user_is_logged_in = false;
            $this->messages[] = "You have been logged out.";     
    }
    
    /**
     * simply return the current state of the user's login
     * @return boolean user's login status
     */
    public function isUserLoggedIn() {
        
        return $this->user_is_logged_in;
        
    }

}

1-minimal/views/register.php
<!-- this is the Simple sexy PHP Login Script. You can find it on http://www.php-login.net ! It's free and open source. -->

<!-- errors & messages --->
<?php

// show negative messages
if ($registration->errors) {
    foreach ($registration->errors as $error) {
        echo $error;    
    }
}

// show positive messages
if ($registration->messages) {
    foreach ($registration->messages as $message) {
        echo $message;
    }
}

?>   

<!-- register form -->
<form method="post" action="register.php" name="registerform">   
    
    <!-- the user name input field uses a HTML5 pattern check -->
    <label for="login_input_username">Username (only letters and numbers, 2 to 64 characters)</label>
    <input id="login_input_username" class="login_input" type="text" pattern="[a-zA-Z0-9]{2,64}" name="user_name" required />
	<br>
    
    <!-- the email input field uses a HTML5 pattern check (3-64 characters) and a HTML5 email type check -->
    <label for="login_input_email">User's email</label>    
    <input id="login_input_email" class="login_input" type="email" pattern="(.*){3,64}" name="user_email" required />
	<br>
    
    <label for="login_input_password_new">Password</label>
    <input id="login_input_password_new" class="login_input" type="password" name="user_password_new" required autocomplete="off" />
	<br>
    
    <label for="login_input_password_repeat">Repeat password</label>
    <input id="login_input_password_repeat" class="login_input" type="password" name="user_password_repeat" required autocomplete="off" />
	<br>
	
    <input type="submit"  name="register" value="Register" />
    
</form>

<!-- backlink -->
<a href="index.php">Back to Login Page</a>

<!-- this is the Simple sexy PHP Login Script. You can find it on http://www.php-login.net ! It's free and open source. -->

1-minimal/views/not_logged_in.php
<!-- this is the Simple sexy PHP Login Script. You can find it on http://www.php-login.net ! It's free and open source. -->

<!-- errors & messages --->

<?php

// show negative messages
if ($login->errors) {
    foreach ($login->errors as $error) {
        echo $error;    
    }
}

// show positive messages
if ($login->messages) {
    foreach ($login->messages as $message) {
        echo $message;
    }
}

?>             

<!-- login form box -->
<form method="post" action="index.php" name="loginform">
    <label for="login_input_username">Username</label>
    <input id="login_input_username" class="login_input" type="text" name="user_name" required />
	<br>
    <label for="login_input_password">Password</label>
    <input id="login_input_password" class="login_input" type="password" name="user_password" autocomplete="off" required />
	<br>
    <input type="submit"  name="login" value="Log in" />
</form>

<a href="register.php">Register new account</a>

<!-- this is the Simple sexy PHP Login Script. You can find it on http://www.php-login.net ! It's free and open source. -->

1-minimal/db_registration.php
<?php

/**
 * class Registration
 * handles the user registration
 * 
 * @author Panique <panique@web.de>
 * @version 1.0
 */
class Registration {

    private     $db_connection              = null;                     // database connection   
    
    private     $user_name                  = "";                       // user's name
    private     $user_email                 = "";                       // user's email
    private     $user_password              = "";                       // user's password (what comes from POST)
    private     $user_password_hash         = "";                       // user's hashed and salted password
    
    public      $registration_successful    = false;

    public      $errors                     = array();                  // collection of error messages
    public      $messages                   = array();                  // collection of success / neutral messages
    
    
    /**
     * the function "__construct()" automatically starts whenever an object of this class is created,
     * you know, when you do "$login = new Login();"
     */    
    public function __construct() {
        
		if (isset($_POST["register"])) {
			
			$this->registerNewUser();
			
		}        
    }

    /**
     * registerNewUser
     * 
     * handles the entire registration process. checks all error possibilities, and creates a new user in the database if
     * everything is fine
     */
    private function registerNewUser() {
        
        if (empty($_POST['user_name'])) {
          
            $this->errors[] = "Empty Username";

        } elseif (empty($_POST['user_password_new']) || empty($_POST['user_password_repeat'])) {
          
            $this->errors[] = "Empty Password";            
            
        } elseif ($_POST['user_password_new'] !== $_POST['user_password_repeat']) {
          
            $this->errors[] = "Password and password repeat are not the same";   
            
        } elseif (strlen($_POST['user_name']) > 64) {
            
            $this->errors[] = "Username cannot be longer than 64 characters";
                        
        } elseif (!preg_match('/^[a-z\d]{2,64}$/i', $_POST['user_name'])) {
            
            $this->errors[] = "Username does not fit the name sheme: only a-Z and numbers are allowed, 2 to 64 characters";
            
        } elseif (empty($_POST['user_email'])) {
            
            $this->errors[] = "Email cannot be empty";
            
        } elseif (strlen($_POST['user_email']) > 64) {
            
            $this->errors[] = "Email cannot be longer than 64 characters";
            
        } elseif (!filter_var($_POST['user_email'], FILTER_VALIDATE_EMAIL)) {
            
            $this->errors[] = "Your email adress is not in a valid email format";
        
        } elseif (!empty($_POST['user_name'])
                  && strlen($_POST['user_name']) <= 64
                  && preg_match('/^[a-z\d]{2,64}$/i', $_POST['user_name'])
                  && !empty($_POST['user_email'])
                  && strlen($_POST['user_email']) <= 64
                  && filter_var($_POST['user_email'], FILTER_VALIDATE_EMAIL)
                  && !empty($_POST['user_password_new']) 
                  && !empty($_POST['user_password_repeat']) 
                  && ($_POST['user_password_new'] === $_POST['user_password_repeat'])) {
            
            // TODO: the above check is redundand, but from a developer's perspective it makes clear
            // what exactly we want to reach to go into this if-block

            // creating a database connection
			try {
			
				$create_db = false;
				
				if(!file_exists(DB_FILENAME))
					$create_db = true;
					
				$this->db_connection = new PDO('sqlite:'.DB_FILENAME);
				
				if($create_db)
					$this->db_connection->exec(file_get_contents('schema.sql'));
				
				$this->db_connection->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
				
				$this->db_connection->setAttribute(PDO::ATTR_EMULATE_PREPARES, false);
				
			} catch (PDOException $e) {
			
				//$this->errors[] = "Sorry, no database connection.";
				$this->errors[] = $e;
				
				return;
			}
			
            // if no connection errors (= working database connection)

			// cut password to 1024 chars to prevent too much calculation
			$this->user_password        = substr($_POST['user_password_new'], 0, 1024);

			/* 
			 * get_salt()
			 * generate random string "salt", a string to "encrypt" the password hash
			 * this is a basic salt, you might replace this with a more advanced function
			 * @see http://en.wikipedia.org/wiki/Salt_(cryptography)
			 */
			function get_salt($length) {

				$options = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789./';
				$salt = '';

				for ($i = 0; $i <= $length; $i ++) {
					$options = str_shuffle ( $options );
					$salt .= $options [rand ( 0, 63 )];
				}
				return $salt;
			}

			// getting the max salt length on your system (usually 123 characters on linux)
			$max_salt = CRYPT_SALT_LENGTH;

			// hard to explain, this part of the upcoming hash string is some kind of parameter, defining
			// the intensity of calculation. we are using the SHA-512 algorithm here, please see
			// @see php.net/manual/en/function.crypt.php
			// for more information.
			$hashing_algorithm = '$6$rounds=5000$';

			//get the longest salt, could set to 22 crypt ignores extra data
			$salt = get_salt($max_salt);

			//append salt data to the password, and crypt using salt, results in a 118 character output
			$this->user_password_hash = crypt($this->user_password, $hashing_algorithm.$salt);

			// check if user already exists
			$statement = $this->db_connection->prepare('SELECT * FROM users WHERE user_name = :username;');
			$statement->execute(array('username' => $_POST['user_name']));
			
			if ($statement->rowCount() == 1) {

				$this->errors[] = "Sorry, that user name is already taken.<br/>Please choose another one.";

			} else {

				// write new users data into database
				$statement = $this->db_connection->prepare("INSERT INTO users (user_name, user_password_hash, user_email) VALUES(:username, :hash, :email);");
				$statement->execute(
					array(
						'username' => $_POST['user_name'],
						'hash' => $this->user_password_hash,
						'email' => $_POST['user_email']
						)
					);

				if ($statement->rowCount() == 1) {

					$this->messages[] = "Your account has been created successfully. You can now log in.";
					$this->registration_successful = true;

				} else {

					$this->errors[] = "Sorry, your registration failed. Please go back and try again.";

				}
			}
            
        } else {
            
            $this->errors[] = "An unknown error occured.";
            
        }
        
    }

}